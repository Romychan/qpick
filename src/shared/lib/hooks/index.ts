export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
