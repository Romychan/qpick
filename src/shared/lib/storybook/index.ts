export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
