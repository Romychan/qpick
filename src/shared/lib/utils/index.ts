export { formatPrice, formatCardNumber, formatExpiryDate } from './formatters';
export { cl } from './className';
