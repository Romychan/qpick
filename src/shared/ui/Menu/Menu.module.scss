.container {
  &.vertical {
    & .list {
      flex-direction: column;
      gap: 0.6667rem;
    }
  }

  &.horizontal {
    & .list {
      flex-direction: row;
      align-items: center;
      gap: 1rem;
    }
  }
}

.list {
  display: flex;
}
