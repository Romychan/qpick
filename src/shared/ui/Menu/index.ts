export { MenuItem } from './MenuItem/MenuItem';
export { Menu } from './Menu';
export type { MenuItemOption } from './types';
