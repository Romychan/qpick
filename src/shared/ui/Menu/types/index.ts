import { ReactNode } from 'react';

/** Base menu item interface */
export interface MenuItemOption {
  /** Route link */
  link: string;
  /** The displayed value of the element */
  label: string;
}

/** Base menu component interface */
export interface MenuBaseProps {
  /**
   * The direction of the menu items
   *
   * @default 'horizontal'
   */
  direction?: 'horizontal' | 'vertical';
  /** Additional class for the component */
  className?: string;
}

/** The type for the menu component without children elements */
export type MenuWithoutChildren = {
  /** List of menu items. If the list is empty, the `children` props will be displayed */
  items: MenuItemOption[];
  /** The content of the component */
  children?: undefined;
} & MenuBaseProps;

/** The type for the menu component with children elements */
export type MenuWithChildren = {
  /** List of menu items. If the list is empty, the `children` props will be displayed */
  items?: undefined;
  /** The content of the component */
  children: ReactNode;
};
