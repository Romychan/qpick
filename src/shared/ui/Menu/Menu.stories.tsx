import type { Meta, StoryObj } from '@storybook/react';

import { withGapDecorator, withRouterDecorator } from '../../lib/storybook';

import { Menu } from './Menu';

const meta = {
  title: 'UI/Menu',
  component: Menu,
  parameters: {
    layout: 'centered',
  },
  decorators: [withRouterDecorator],
  args: {
    items: [
      {
        link: '/link1',
        label: 'Link #1',
      },
      {
        link: '/link2',
        label: 'Link #2',
      },
      {
        link: '/link3',
        label: 'Link #3',
      },
    ],
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Menu>;

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to default */
export const Default: Story = {};

/**
 * The `direction` property is used to change the direction of positions menu.
 * - `horizontal` is used to direction menu items horizontally
 * - `vertical` is used to direction menu items vertically
 */
export const Variants: Story = {
  decorators: [withGapDecorator],
  render: (args) => {
    return (
      <>
        <Menu {...args} direction="horizontal" />
        <Menu {...args} direction="vertical" />
      </>
    );
  },
};

/** To display custom elements, make the list of menu items empty and pass custom elements as children */
export const CustomElements: Story = {
  decorators: [withGapDecorator],
  render: (args) => {
    return (
      <>
        <Menu {...args} items={undefined}>
          <p>Element #1</p>
          <p>Element #2</p>
          <p>Element #3</p>
        </Menu>
      </>
    );
  },
};
