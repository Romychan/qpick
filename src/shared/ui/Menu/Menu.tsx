import { cl } from '../../lib/utils';

import { MenuItem } from './MenuItem/MenuItem';
import { MenuBaseProps, MenuWithChildren, MenuWithoutChildren } from './types';
import styles from './Menu.module.scss';

type IMenuProps = (MenuWithChildren | MenuWithoutChildren) & MenuBaseProps;

/** This component is used to display the menu list and navigation */
export const Menu = ({
  direction = 'horizontal',
  className,
  items,
  children,
}: IMenuProps) => {
  return (
    <nav className={cl(styles.container, styles[direction], className)}>
      <ul className={styles.list}>
        {children && !items
          ? children
          : items?.map(({ link, label }) => (
              <MenuItem key={link} link={link}>
                {label}
              </MenuItem>
            ))}
      </ul>
    </nav>
  );
};
