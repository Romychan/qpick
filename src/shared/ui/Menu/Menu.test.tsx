import { screen } from '@testing-library/react';

import { renderWithRouter } from '../../lib/tests';

import { Menu } from './Menu';

const MOCK_MENU = [
  {
    link: '/link1',
    label: 'Link #1',
  },
  {
    link: '/link2',
    label: 'Link #2',
  },
  {
    link: '/link3',
    label: 'Link #3',
  },
];

describe('Menu', () => {
  it('should render correctly', () => {
    renderWithRouter(<Menu items={MOCK_MENU} />);

    const itemsElement = screen.getAllByTestId('menu-item');

    expect(itemsElement).toHaveLength(3);
  });

  it('should render correctly with direction', () => {
    renderWithRouter(<Menu direction="horizontal" items={MOCK_MENU} />);

    const navigationElement = screen.getByRole('navigation');

    expect(navigationElement.getAttribute('class')).toMatch('horizontal');
    expect(navigationElement.getAttribute('class')).not.toMatch('vertical');
  });

  it('should render correctly if items is empty, but there are children', () => {
    renderWithRouter(
      <Menu>
        <button>Test</button>
      </Menu>,
    );

    const itemElement = screen.queryByTestId('menu-item');
    const buttonElement = screen.getByRole('button', { name: 'Test' });

    expect(buttonElement).toBeInTheDocument();
    expect(itemElement).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const container = renderWithRouter(<Menu items={MOCK_MENU} />);

    expect(container).toMatchSnapshot();
  });
});
