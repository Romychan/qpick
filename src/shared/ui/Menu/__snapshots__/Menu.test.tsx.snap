// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="_container_ca994d _horizontal_ca994d"
      >
        <ul
          class="_list_ca994d"
        >
          <li
            class="_container_cf770d"
            data-testid="menu-item"
          >
            <a
              class="_link_cf770d"
              href="/link1"
            >
              Link #1
            </a>
          </li>
          <li
            class="_container_cf770d"
            data-testid="menu-item"
          >
            <a
              class="_link_cf770d"
              href="/link2"
            >
              Link #2
            </a>
          </li>
          <li
            class="_container_cf770d"
            data-testid="menu-item"
          >
            <a
              class="_link_cf770d"
              href="/link3"
            >
              Link #3
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="_container_ca994d _horizontal_ca994d"
    >
      <ul
        class="_list_ca994d"
      >
        <li
          class="_container_cf770d"
          data-testid="menu-item"
        >
          <a
            class="_link_cf770d"
            href="/link1"
          >
            Link #1
          </a>
        </li>
        <li
          class="_container_cf770d"
          data-testid="menu-item"
        >
          <a
            class="_link_cf770d"
            href="/link2"
          >
            Link #2
          </a>
        </li>
        <li
          class="_container_cf770d"
          data-testid="menu-item"
        >
          <a
            class="_link_cf770d"
            href="/link3"
          >
            Link #3
          </a>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
